### DISCUSSION TITLE
Struct embedding init syntax

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#7568] [General]
Currently our syntax is like Gos syntax:
```v
Programmer{
    Person: Person{
        name: 'Daniel'
    }
}
```

We have two proposals for a new syntax:

1.
```v
Programmer{
    name: 'Daniel'
}
```
2.
```v
Programmer{
    Person{
        name: 'Daniel'
    }
}
```

There are some problems with this and I want to clarify it to decide how it should be implemented.

1.
If programmer looks like this:
```v
struct Programmer{
    Person
    Computer
}
```
and both, Person and Computer have a field called `name`, you can't use the syntactic sugar anymore.

2.
if programmer would look like this:
```v
Programmer{
    mod.Person
}
```
and you initialize is like this:
```v
Programmer{
    mod.Person{
        name: 'Daniel'
    }
}
```
accessing this embed like this
```v
programmer.Person.name
```
could get confusing. Why is the name of the field `Person`? and not `mod_Person` or something similar. In other words, it's a bit implicit and looks like magic. The user have to know it.
Whereas 
```v
Programmer{
    Person: mod.Person{}
}
```
The user also have to know to initialize with `Person` as the key, but `programmer.Person` is no magic anymore.

**Comment by medvednikov** [comment]
> and both, Person and Computer have a field called name

we discussed this, it should result in a compilation error

this is the desired syntax:

```
Programmer{
    name: 'Daniel'
}
```

**Reply by danieldaeschle** [reply]
and the alternative is `Person: Person{}`?

**Reply by medvednikov** [reply]
alternative for what?

**Reply by danieldaeschle** [reply]
```v
struct Programmer{
    Person
    name string
}
struct Person{
    name string
}

prog := Programmer{
    name 'Daniel' // this initializes field of Programmer and not of Person
}

// alternative to initialize the embedded struct would be:
prog := Programmer{
    Person: Person{
        name: 'Daniel'
    }
}
```

**Reply by medvednikov** [reply]
Again, duplications always lead to a compilation error.

**Reply by danieldaeschle** [reply]
But in Go it works: https://play.golang.com/p/hbknjuPa19n

**Reply by medvednikov** [reply]
it doesn't 

	fmt.Println(a.name)

./prog.go:25:15: ambiguous selector a.name

**Reply by danieldaeschle** [reply]
You showed accessing the field. This is already covered. But this discussion is about the embedding itself. In go you can access it using `fmt.Println(a.Person.name)`.
Let me show you more examples. I think we are talking about different things.

(in other words: the embedding doesn't result in a compilation error but the accessing of duplicated fields)

**Reply by medvednikov** [reply]
In V a duplicate field must result in an error right away.

**Reply by danieldaeschle** [reply]
Let us begin with something easy:
Should both syntax be possible?
```v
prog := Programmer{
    Person: Person{
        name: 'Daniel'
    }
}
```
and
```v
prog := Programmer{
    name: 'Daniel'
}
```

**Reply by medvednikov** [reply]
I'm not 100% sure, I think just normal syntax should be enough (name:'')

**Reply by medvednikov** [reply]
Just to keep it simple and "one way".

If we see there's a need to specify the object (e.g. `Programmer{ Person: person }`), we can add that later.

**Reply by danieldaeschle** [reply]
Ok, thanks. I will remove that functionality then.

**Comment by JalonSolov** [comment]
Not an issue, because it would have already failed compilation due to having `name` in both structs.

**Reply by danieldaeschle** [reply]
No, it's possible to have two embeddings with same fields. But accessing them is different then. https://play.golang.com/p/hbknjuPa19n

**Reply by medvednikov** [reply]
Again, in V duplicate fields are not allowed.

**Comment by JalonSolov** [comment]
It doesn't matter what Go allows.  It has been declared that V will _NOT_ allow duplicate field names in embedded structures.

There is just too much chance of referencing the wrong one if multiples are allowed.

This is simply a case of V trying to be better than Go.

**Reply by danieldaeschle** [reply]
Okay, now I understand. Then the currently implemented functionality should be cleaned up. What about methods?

Should `x.Person.name` still be allowed?
What is if `x` implements the same field or method than `Person`? Is this field / method still accessable using `x.Person.method()` or do we throw away that syntax completely?

**Comment by JalonSolov** [comment]
Hmm...

My first thought was `no, x.Person.name` would not be allowed, because the fields of the embedded struct now look like fields in the struct they are embedded into.
```v
struct Person {
    name string
    age int
}

struct Programmer {
    Person
    computer_languages []string
}

jalon := Programmer{name: 'Jalon', age: 62, computer_languages: ['C', 'Go', 'Java', 'V' /* and lots of others */]}
```

This can be odd, though, if you import a module and embed a struct that has receiver functions declared in the module.  Do those receiver functions get changed so that 'Programmer' is now the receiver?  I mean, that's pretty much how it would have to work, but... that would be really interesting to figure out what already existed unless you read the module source.

**Comment by danieldaeschle** [comment]
To clarify this finally. Current state:
```v
struct Person {
    name string
    age int
}
struct Pet {
    name string
    color Color
}
```
1.
```v
//> error: cannot embed struct `Pet`, it shares the same field `name` as struct `Person`
struct Programmer {
    Person
    Pet
}
```
Is this correct? Should that behave the same with methods? If yes, it could easily get difficult for the user to embed two structs. Very often structs share the same field names.

2. What happens if the upper struct overwrites the embedded field name? (methods are still accessable using `programmer.Person.method()`.
```v
type ProgrammerName = string
struct Programmer {
    Person
    name ProgrammerName
}

//> this initializes `Programmer`s `name`, not `Person`s `name`
p := Programmer{
    name: 'test'
}
```
But how to initialize `Person`s `name` then?
Would this be the solution? It looks inconsistent imo.
```v
p := Programmer{
    name: 'test'
    age: 18
    Person: Person{
        name: 'test'
    }
}
```
All in all it looks really strange and i need a bit more information how to implement proper embedding. Please answer this two questions and I will ask more if I get some scenarios which are still weird to me.

**Reply by medvednikov** [reply]
> Is this correct? Should that behave the same with methods?
yes

number 2 should result in a compilation error for the same reason

Programmer.name and Person.name conflict

**Reply by danieldaeschle** [reply]
But this should not happen with methods, right?

**Reply by spaceface777** [reply]
not directly related, but imho, conflicts in an embedded struct's private fields/methods shouldn't cause compilation errors (if embedding a struct imported from another module), since the user can't access those fields/methods anyways.

**Reply by danieldaeschle** [reply]
@danieldaeschle and what about initalization?

**Comment by JalonSolov** [comment]
And what about receiver methods?

For example, if there was both
```v
fn (p Person) set_name(name string) {
```
and
```v
fn (p Pet) set_name(name string) {
```
Should that fail with a conflict?  Probably.  But how do you know if a struct in a module you imported has that receiver function ahead of time?  Maybe it's private...

**Reply by danieldaeschle** [reply]
You could check it in checker. But we have that scenario in vweb. I don't know if we should forbid that.

**Comment by Ben-Fields** [comment]
My 2 cents:
Disallowing name conflicts is too restrictive. Sometimes you do not have naming control.
Given:
```v
struct Person {
    name string
    age int
}
struct Pet {
    name string
    color Color
}
fn (p Person) set_name(name string) { /*fn body*/ }
fn (p Pet) set_name(name string) { /*fn body*/ }
```
The following should be valid/invalid:
```v
struct Programmer {   // valid
    Person
    Pet {
        name: 'George Boole'
    }
}
...
me := Programmer {    // valid
    age: 100
    Person {
        name: 'Me'
    }
}
println(me.age)              // valid (output: '100')
println(me.name)             // invalid (compiler error: ambiguous struct member)
println(me.Person.name)      // valid (output: 'Me')
me.set_name('Joe')           // invalid (compiler error: ambiguous struct receiver method)
me.Person.set_name('Joe')    // valid
```
Keep the rule set simple and without removing power. (The rules for members and receiver methods are the same).
Keep in mind the verbosity of `me.Person.name` only occurs with naming conflicts.
The compiler only searches the current scope (naming conflicts only happen if they are in the module and/or public and imported).
One syntax. No `Person : Person`; It confuses semantics.
Some of this is repeated above, but I hope my outside thoughts are helpful.

**Reply by Ben-Fields** [reply]
From the original post:
> `programmer.Person.name` could get confusing. Why is the name of the field Person? and not mod_Person or something similar. In other words, it's a bit implicit and looks like magic. The user have to know it.

`Person`, in this context, is a specifier solely to remove ambiguity, not a field. It is similar to how you must specify the module name since two modules could have an identically-named function. The only difference is case style, which I think is neat here since it stands out as an ambiguity specifier.

**Comment by Ben-Fields** [comment]
I would also propose to make the `union` syntax identical to the (embedded) `struct` syntax once this is decided.  #7221
(I meant the syntax format should be the same, not that the union should be the same as a struct. See reply below).

**Comment by JalonSolov** [comment]
union is defined in many other languages, and is expected to work the way it does now.  That is highly unlikely to change.

**Reply by Ben-Fields** [reply]
I agree the behavior should be unchanged. I meant syntactically writing unions should be as easy as structs.
Given:
```v
struct Person {
    name string
    age int
}
struct Pet {
    name string
    color Color
}
```
Both of these should be the same except for memory layout (the struct members are separate, while the union members overlap):
```v
struct Programmer {
    Person
    Pet
}
union UProgrammer {
    Person
    Pet
}
...
me := Programmer {
    Person {
        name: 'Me'
    }
    Pet {
        name: 'Boole'
    }
}
my_thought := UProgrammer {
    Person {            // only one initializer since it's a union
        name: 'Joe'
        age: 101
    }
}
```
And both can be accessed syntactically as I described above (or in some other identical manner). The only difference is that the union members read/write from the same space.

Currently, you have to do:
```v
struct Comp {
	b byte
	g byte
	r byte
	a byte
}
union Uni {
	value u32
	comp Comp
}
...
c := Comp {
    r: 2
    g: 2
    b: 4
    a: 5
}
uni := Uni {
    comp: c
}
println(uni.comp.r)
```

**Comment by sanathusk** [comment]
How about using something like a spread operator (...).
The programmer can control the embedded struct fields needs to be spread (propagated to enveloping struct).
Case 1: When you want to merge fields of embedded struct into enveloping struct:
```
struct Person {
    name string
    age int
}
struct Programmer {
 ...Person
languages_known : [] string
}
```
Case 2: When you want to preserve the embedded struct fields:

```
struct Pet  {
name  string
age int
}

struct Person {
    name string
    age int
    pet Pet
}
```
now, person.name is different from pet.name & you could access pet name using person.pet.name

**Reply by dumblob** [reply]
This looks really good to me @sanathusk, but unfortunately violates the "there is only one way of doing a thing" which I admittedly prefer over having **any** syntactical option how to use embedded structs.

Admittedly I've not yet seen any persuasive argument why not to enforce the syntax `programmer.Person.name` everywhere (it's explicit, simple and I don't see any inconsitency - at the declaration it's clear what'll conflict and what not and that struct's members can't be named as the types being embedded into the struct and that accessing these embedded members goes always through the same-looking syntax as struct declaration).

Maybe there is fear of potential syntactical incompatibility with future additions in the language (not `union`s as they **are** compatible with this syntax?

**Reply by JalonSolov** [reply]
It's pretty simple to just NOT embed, then you can use that syntax as you like.
```v
struct Person {
mut:
    name string
    age int
}

struct Pet {
mut:
    name string
    age int
}

struct Programmer {
mut:
    person Person
    pet Pet
}

mut p := Programmer{}
p.person.name = 'Bob'
p.pet.name = 'Fifi'

println(p)
```
output is
```
Programmer{
    person: Person{
        name: 'Bob'
        age: 0
    }
    pet: Pet{
        name: 'Fifi'
        age: 0
    }
}
```
struct embedding is for when you DON'T want the extra layer of names.

**Reply by JalonSolov** [reply]
Ok, not just to not have an extra layer of names.  What struct embedding is really useful for is if you have a lot of structs with some shared fields, why not extract the shared fields to a separate struct that you embed in the rest, so you don't have to copy/paste/re-type in every single one.

For example, UI code.  Every widget likely has width, height, x, y.  So
```v
struct WidgetBase {
    width int
    height int
    x int
    y int
}

struct ListBox {
    WidgetBase
    // listbox specific fields
}

struct TextBox {
    WidgetBase
    // textbox specific fields
}

// 30+ other widgets, all with WidgetBase embedded
```
In this case, it would be very tedious to have to type `widget.WidgetBase.width = 200` etc., etc.  when `widget.width = 200` works, and is what most people would expect.

**Reply by dumblob** [reply]
>struct embedding is for when you DON'T want the extra layer of names

That makes sense, but I don't find this argument enough to defend the sole existence of struct embedding (note I'm not talking about unions here). Any other reasons why *V* wants to support embedding then? Even the use case *making already existing methods work seamlessly with your very own structs* (see https://github.com/vlang/v/issues/2025#issuecomment-533903276 as demonstrated by the pseudo-code in https://github.com/vlang/v/issues/2025#issuecomment-536203283 ) doesn't seem to apply here.

I'm really not sure whether all the hassle with embedding and it's complicated semantics (padding rules, etc.) and issues with future language extensions is a good idea.

Thanks to `vfmt` and other tools it seems dead easy to transform between the non-embedded syntax and the embedded syntax (without any difference in readability nor number of lines of code) which per se clearly violates the rule *there is only one way od doing a thing*.

**Reply by dumblob** [reply]
@JalonSolov hm, so it's not about embedding (padding, etc. with complicated semantics), but just a syntax sugar. Well, that's much easier to define and use then. If we switch our thinking from embedding to just a syntax sugar for structs with many common building blocks, then we could come up with simpler solution to all the open questions about "embedding" by simply avoiding them at all.

**Reply by Ben-Fields** [reply]
On the surface, nested structs seem to have a different meaning than embedded structs. Ex. `Pet` was embedded in `Programmer` for the sake of example above, but I would write it nested in a more realistic scenario. (programmer.color doesn't make sense, but programmer.pet.color does).

**Comment by JalonSolov** [comment]
embedded - `fixed firmly and deeply in a surrounding mass; implanted. "a gold ring with nine embedded stones"`
nested - `placed or stored one inside the other. "a set of three nested tables"`

structs are `embedded` in V by the actual meaning of the word.  If some other language uses that word to mean something different, then they are using the word incorrectly.

**Reply by Ben-Fields** [reply]
Thanks for that. Now I get what @dumblob was saying:
> so it's not about embedding (padding, etc. with complicated semantics), but just a syntax sugar.
